### DISCUSSION TITLE
Compiling into a statically-linked DLL

### DISCUSSION THREAD

**Discussion by richardg867** [FluidSynth/fluidsynth#1001] [Ideas]
Some projects, like [86Box/86Box](https://github.com/86Box/86Box) (which I help maintain) and [coelckers/gzdoom](https://github.com/coelckers/gzdoom), ship their Windows builds with an ancient version of FluidSynth compiled into a single statically-linked DLL, to avoid the burden of shipping separate DLLs for each of FluidSynth's dependencies (libgcc, glib and so on). That DLL was built many years ago by an unknown person out of an unknown FluidSynth 1.x codebase through unknown means. Therefore, it would be interesting to have an officially-supported way to compile a modern version of FluidSynth into a single DLL, or a single dynamic library on other platforms (though Windows is my main concern). I have attempted to produce a single DLL myself, either through modifying the CMake definitions or running the static library through `dllwrap`, but I haven't had any luck.

**Comment by derselbst** [comment]
> burden of shipping separate DLLs for each of FluidSynth's dependencies (libgcc, glib and so on)

What exactly makes this a burden? You just have multiple DLLs and ship them in the same directory. That's it, isn't it? If there's a bug in fluidsynth, you replace the fluidsynth.dll. If there's a bug in libsndfile, you replace libsndfile.dll, etc...

IMO, the real burden is getting such a customized build working (and maintainable!) which you're asking for. Esp. because glib cannot be statically compiled on Windows, IIRC.

**Reply by OBattler** [reply]
Hello!

I'm the maintainer of 86Box, and I can explain why requiring all those .dll's is a problem - there's no easy way to keep track of which exactly are needed, which can potentially cause a mess if a new independency is reduced and/or an existing one removed. We simply don't have resources to manually check for the dependencies on every build.

Edit: To further explain the problem - the dependency dll's have their own dependencies as well, which causes there to be a whole tree of dependencies.

Edit #2: To clarify - I'm using Msys2.

**Comment by ghost** [comment]
This is a good idea and I have implemented just now:
https://github.com/3119369616/fluidsynth/releases/tag/v2.2.7
This binary is 64-Bit and can be used by MinGW-w64 or CLang (x86_64-pc-windows-gnu toolchain) compiler on Windows. 
(I'm not sure whether it can be used by MSVC.)
NOTE: Just append the argument `--static` into GCC so that you can decrease the count of depended DLLs such as libwinpthread-1.dll, etc.
Please contact me immediately if you find any errors.   (^_^)

**Reply by OBattler** [reply]
Would a 32-bit version be possible to do?
